powerauth-restful-security-spring/src/main/java/io/getlime/security/powerauth/rest/api/spring/controller/v3/ActivationController.java

```diff
@@ -0,0 +1,194 @@
+
+        if (eciesEncryption == null) {
+            throw new PowerAuthAuthenticationException("ECIES object is missing");
+        }
+        try {
+
+            switch (request.getType()) {
+                // Regular activation which uses "code" identity attribute
+                case CODE:
+                    // Extract data from request and encryption object
+                    String activationCode = request.getIdentityAttributes().get("code");
+                    String applicationKey = eciesEncryption.getApplicationKey();
+                    EciesEncryptedRequest activationData = request.getActivationData();
+                    String ephemeralPublicKey = activationData.getEphemeralPublicKey();
+                    String encryptedData = activationData.getEncryptedData();
+                    String mac = activationData.getMac();
+
+                    // Call PrepareActivation SOAP method on PA server
+                    PrepareActivationResponse response = powerAuthClient.prepareActivation(activationCode, applicationKey, ephemeralPublicKey, encryptedData, mac);
+
+                    // Prepare encrypted response object for layer 2
+                    EciesEncryptedResponse encryptedResponseL2 = new EciesEncryptedResponse();
+                    encryptedResponseL2.setEncryptedData(response.getEncryptedData());
+                    encryptedResponseL2.setMac(response.getMac());
+
+                    // The response is encrypted once more before sent to client using ResponseBodyAdvice
+                    ActivationLayer1Response responseL1 = new ActivationLayer1Response();
+                    responseL1.setActivationData(encryptedResponseL2);
+                    return responseL1;
+
+                // Custom activation
+                case CUSTOM:
+                    throw new IllegalStateException("Not implemented yet");
+
+                default:
+                    throw new PowerAuthAuthenticationException("Unsupported activation type: "+request.getType());
+            }
+        } catch (Exception ex) {
+            logger.warn("Creating PowerAuth activation failed.", ex);
+            throw new PowerAuthAuthenticationException(ex.getMessage());
+        }
+    }
+
+    /**
+     * Get activation status.
+     * @param request PowerAuth RESTful request with {@link ActivationStatusRequest} payload.
+     * @return PowerAuth RESTful response with {@link ActivationStatusResponse} payload.
+     * @throws PowerAuthActivationException In case request fails.
+     */
+    @RequestMapping(value = "status", method = RequestMethod.POST)
+    public ObjectResponse<ActivationStatusResponse> getActivationStatus(@RequestBody ObjectRequest<ActivationStatusRequest> request)
+            throws PowerAuthActivationException {
+        try {
+            String activationId = request.getRequestObject().getActivationId();
```

Comment:
Maybe we could validate the request here a bit, to avoid NPE and weird activation ID values...